Util/channel/src/main/java/io/deephaven/util/channel/DataInstructionsProviderLoader.java

```diff
@@ -0,0 +1,82 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.util.channel;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.net.URI;
+import java.util.*;
+
+/**
+ * A service loader class for loading {@link DataInstructionsProviderPlugin} implementations at runtime which provide
+ * {@link DataInstructionsProviderLoader} implementations for different URI paths.
+ */
+public final class DataInstructionsProviderLoader {
+    /**
+     * A weakly held cache of {@link DataInstructionsProviderLoader} instances keyed by the property collection.
+     */
+    private static final WeakHashMap<Map<String, String>, DataInstructionsProviderLoader> instances =
+            new WeakHashMap<>();
+
+    /**
+     * Get a {@link DataInstructionsProviderLoader} instance for the given property collection.
+     *
+     * @param properties The property collection.
+     * @return A {@link DataInstructionsProviderLoader} instance.
+     */
+    public static DataInstructionsProviderLoader getInstance(final Map<String, String> properties) {
+        return instances.computeIfAbsent(properties, DataInstructionsProviderLoader::new);
+    }
+
+    /**
+     * The properties collection for this instance.
+     */
+    private final Map<String, String> properties;
+
+    /**
+     * The list of plugins loaded by the {@link ServiceLoader}.
+     */
+    private final List<DataInstructionsProviderPlugin> providers;
+
+    /**
+     * A weakly held cache of {@link DataInstructionsProviderPlugin} instances keyed by the URI.
+     */
+    private final WeakHashMap<URI, Object> cache;
```

Comment:
I'm not sure how useful this will really be. It might be better to skip caching.